streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java

```diff
@@ -293,27 +295,33 @@ private <K, V> void handleException(final ProductionExceptionHandler.Serializati
                                         final Long timestamp,
                                         final String processorNodeId,
                                         final InternalProcessorContext<Void, Void> context,
-                                        final Exception exception) {
+                                        final RuntimeException serializationException) {
+        log.debug(String.format("Error serializing record for topic %s", topic), serializationException);
+
+        final DefaultErrorHandlerContext errorHandlerContext = new DefaultErrorHandlerContext(
+            null, // only required to pass for DeserializationExceptionHandler
+            context.recordContext().topic(),
+            context.recordContext().partition(),
+            context.recordContext().offset(),
+            context.recordContext().headers(),
+            processorNodeId,
+            taskId
+        );
         final ProducerRecord<K, V> record = new ProducerRecord<>(topic, partition, timestamp, key, value, headers);
-        final ProductionExceptionHandlerResponse response;
-
-        log.debug(String.format("Error serializing record to topic %s", topic), exception);
 
+        final ProductionExceptionHandlerResponse response;
         try {
-            final DefaultErrorHandlerContext errorHandlerContext = new DefaultErrorHandlerContext(
-                null, // only required to pass for DeserializationExceptionHandler
-                context.recordContext().topic(),
-                context.recordContext().partition(),
-                context.recordContext().offset(),
-                context.recordContext().headers(),
-                processorNodeId,
-                taskId
+            response = productionExceptionHandler.handleSerializationException(errorHandlerContext, record, serializationException, origin);
+        } catch (final RuntimeException fatalUserException) {
+            log.error(
+                String.format(
+                    "Production error callback failed after serialization error for record %s: %s",
+                    origin.toString().toLowerCase(Locale.ROOT),
+                    errorHandlerContext
+                ),
+                serializationException
             );
-            response = productionExceptionHandler.handleSerializationException(errorHandlerContext, record, exception, origin);
-        } catch (final Exception e) {
-            log.error("Fatal when handling serialization exception", e);
-            recordSendError(topic, e, null, context, processorNodeId);
-            return;
+            throw new FailedProcessingException("Fatal user code error in production error callback", fatalUserException);
```

Comment:
fix: inside `send()` is seem incorrect to call `recordSendError()` which we only use in from the `producer.send(..., Calback)` which is called async. Here inside `RecordCollector.send()` (before we hit `producer.send()`) can should rather throw directly